### PR TITLE
chore(deploy): add DeployAndRenderEnvironment method

### PR DESCRIPTION
**Pull request by efekarakus**
The "deploy" pkg can now create an environment stack and render the
stack updates to an io.Writer concurrently.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: internal/pkg/aws/cloudformation/cloudformation.go [hunk 1]
```diff
@@ -59,32 +59,42 @@ func (c *CloudFormation) Create(stack *Stack) error {
 	if status.requiresCleanup() {
 		// If the stack exists, but failed to create, we'll clean it up and then re-create it.
 		if err := c.Delete(stack.Name); err != nil {
-			return fmt.Errorf("cleanup previously failed stack %s: %w", stack.Name, err)
+			return "", fmt.Errorf("cleanup previously failed stack %s: %w", stack.Name, err)
```
Comment (huanjani) [inline internal/pkg/aws/cloudformation/cloudformation.go]: tiny nit: "cleanup" should be two words here, "clean up"

File: internal/pkg/aws/cloudformation/cloudformation.go [hunk 2]
```diff
@@ -59,32 +59,42 @@ func (c *CloudFormation) Create(stack *Stack) error {
...
+		return nil, err
+	}
+	return out, nil
 }
 
 // WaitForCreate blocks until the stack is created or until the max attempt window expires.
-func (c *CloudFormation) WaitForCreate(stackName string) error {
-	err := c.client.WaitUntilStackCreateCompleteWithContext(context.Background(), &cloudformation.DescribeStacksInput{
+func (c *CloudFormation) WaitForCreate(ctx context.Context, stackName string) error {
+	err := c.client.WaitUntilStackCreateCompleteWithContext(ctx, &cloudformation.DescribeStacksInput{
```
Comment (huanjani) [inline internal/pkg/aws/cloudformation/cloudformation.go:97]: This method name (aka W37t) wins a prize for length!  😂

**Review by huanjani — APPROVED** [review]
Looks good to me!

File: internal/pkg/deploy/cloudformation/cloudformation.go [hunk 3]
```diff
@@ -73,6 +84,19 @@ func New(sess *session.Session) CloudFormation {
 	}
 }
 
+// ErrorEvents returns the list of Cloudformation Resource Events, filtered by failures and erros.
```
Comment (iamhopaul123) [inline internal/pkg/deploy/cloudformation/cloudformation.go]: nit:
```suggestion
// ErrorEvents returns the list of CloudFormation Resource Events, filtered by failures and errors.
```

File: internal/pkg/deploy/cloudformation/cloudformation.go [hunk 4]
```diff
@@ -102,6 +126,54 @@ func (cf CloudFormation) streamResourceEvents(done <-chan struct{}, events chan
...
+	// When the waiter exits, the waitCtx is canceled which results in the streamer and renderer to exit.
+	// When the streamer exits, the group ctx is canceled resulting in the waiter to exit and canceling the renderer.
+	// When the renderer exits, it's the same flow as the streamer.
+	g, ctx := errgroup.WithContext(waitCtx)
+	g.Go(func() error {
+		defer cancelWait()
+		return in.waitForStack(ctx, in.stackName)
+	})
+	g.Go(func() error {
+		return stream.Stream(waitCtx, streamer)
```
Comment (iamhopaul123) [inline internal/pkg/deploy/cloudformation/cloudformation.go:169]: Do we need to mock this one for testing?

**Review by iamhopaul123 — APPROVED** [review]
What a magic! LGTM just a nit and a question.

File: internal/pkg/deploy/cloudformation/cloudformation.go (hunk 4 above)
Comment (efekarakus) [reply internal/pkg/deploy/cloudformation/cloudformation.go:169]: It's actually tested! See the test case: ["bubbles up streamer error and cancels waiter and renderer on streamer error"](https://github.com/aws/copilot-cli/pull/1793/files#diff-ff51f4486199743cf7adf97941b73ab33d0538bb6cd9e0db2fdc3aab2465d33eR102)

**Pull request merged**